src/workerd/io/actor-cache-test.c++

```diff
@@ -1442,76 +1418,122 @@ KJ_TEST("ActorCache read retry on flush containing only puts") {
-  // we actually expect an exception.
-  auto promise = test.get("foo").get<kj::Promise<kj::Maybe<kj::String>>>();
+  (void)expectUncached(test.get("corge"));
+  auto promise = expectUncached(test.get("foo"));
   test.put("bar", "456");
   test.delete_("baz");
 
-  // Expect the get, but don't resolve yet.
-  auto mockGet = mockStorage->expectCall("get", ws)
-      .withParams(CAPNP(key = "foo"));
+  // Expect the get, but cancel the promise before we finish it.
+  mockStorage->expectCall("getMultiple", ws)
+      .withParams(CAPNP(keys = ["corge", "foo"]), "stream"_kj)
+      .useCallback("stream", [&ws, promise = kj::mv(promise)](MockClient stream) mutable {
+    promise = nullptr;
+    stream.call("end", CAPNP()).expectReturns(CAPNP(), ws);
+  }).thenReturn(CAPNP());
 
-  // The transaction won't write anything until the read completes.
+  // The transaction proceeds.
   auto mockTxn = mockStorage->expectCall("txn", ws).returnMock("transaction");
-  mockTxn->expectNoActivity(ws);
-
-  // Fail out the read with non-disconnect.
-  kj::mv(mockGet).thenThrow(KJ_EXCEPTION(FAILED, "read failed"));
-
-  // The read propagates the error.
-  KJ_EXPECT_THROW_MESSAGE("read failed", promise.wait(ws));
-
-  // The read is NOT retried, so expect the transaction to run now.
   mockTxn->expectCall("delete", ws)
       .withParams(CAPNP(keys = ["baz"]))
       .thenReturn(CAPNP(numDeleted = 0));
   mockTxn->expectCall("put", ws)
       .withParams(CAPNP(entries = [(key = "bar", value = "456")]))
       .thenReturn(CAPNP());
+
   mockTxn->expectCall("commit", ws).thenReturn(CAPNP());
   mockTxn->expectDropped(ws);
 
-  // The read is NOT retried.
-  mockStorage->expectNoActivity(ws);
+  // Since we once asked for these keys, they are now cached even though we dropped the promises.
+  KJ_EXPECT(expectCached(test.get("corge")) == kj::none);
+  KJ_EXPECT(expectCached(test.get("foo")) == kj::none);
 }
 
-KJ_TEST("ActorCache read cancel") {
+KJ_TEST("ActorCache read overwrite") {
   ActorCacheTest test;
   auto& ws = test.ws;
   auto& mockStorage = test.mockStorage;
 
-  auto promise = expectUncached(test.get("foo"));
-  test.put("bar", "456");
-  test.delete_("baz");
+  // Make some gets but overwrite them in the cache with puts.
+  auto promise1 = expectUncached(test.get("foo"));
+  auto promise2 = expectUncached(test.get("bar"));
+  (void)expectUncached(test.get("baz"));
 
-  // Expect the get, but intentionally don't resolve it.
-  auto mockGet = mockStorage->expectCall("get", ws)
-      .withParams(CAPNP(key = "foo"));
+  test.put("foo", "456");
+  test.put("bar", "789");
+  test.put("baz", "123");
 
-  // The transaction won't write anything until the read completes.
-  auto mockTxn = mockStorage->expectCall("txn", ws).returnMock("transaction");
-  mockTxn->expectNoActivity(ws);
+  // Since we still have the promise for foo and bar, we do send a get for them. But baz is not in
```

Comment:
This can be phrased better. When we drop the promise, we drop our `GetWaiter`, which owns a strong reference to the Entry. `~GetWaiter()` will remove the Entry for `baz` from the `dirtyList`, but not from `currentValues`. When we do a `put("baz", "123")` after this, we add the `newEntry` to `dirtyList` and it gets flushed as a put. There is no `Entry` to get from storage since we removed it from the dirtyList earlier.

Reply:
https://github.com/cloudflare/workerd/pull/1916/commits/2089ca41f41da23bd40f198bea269ef2d7ebba38